cdap-watchdog/src/main/java/co/cask/cdap/logging/save/KafkaOffsetFinder.java

```diff
@@ -0,0 +1,284 @@
+                         String topic, int partition) {
+      this.offsetQueue = offsetQueue;
+      this.targetTimestamp = targetTimestamp;
+      this.topic = topic;
+      this.partition = partition;
+      this.brokerInfo = brokerService.getLeader(topic, partition);
+      this.simpleConsumer
+        = new kafka.javaapi.consumer.SimpleConsumer(brokerInfo.getHost(), brokerInfo.getPort(),
+                                                    SO_TIMEOUT, FETCH_SIZE, "simple-kafka-client");
+      this.minOffset = getBoundaryOffset(EARLIEST);
+      // Latest offset returned is 1 larger than the latest received message's offset
+      this.maxOffset = getBoundaryOffset(LATEST) - 1;
+    }
+
+    @Override
+    // Use binary search to find the offset of the message with the timestamp matching targetTimestamp.
+    // Return the offset to fetch the next message until the matching message is found.
+    public long onReceived(Iterator<FetchedMessage> messages) {
+      if (!messages.hasNext()) {
+        return EARLIEST;
+      }
+      FetchedMessage message = messages.next();
+      long offset = message.getOffset();
+      ILoggingEvent event;
+      try {
+        event = serializer.fromBytes(message.getPayload());
+      } catch (IOException e) {
+        LOG.warn("Message with offset {} in topic {} partition {} is ignored because of failure to decode.",
+                 offset, topic, partition, e);
+        return message.getNextOffset();
+      }
+      LOG.trace("Received event {} for topic {} partition {}", event, topic, partition);
+      long timeStamp = event.getTimeStamp();
+      if (timeStamp == targetTimestamp) {
+        if (offsetQueue.isEmpty()) {
+          offsetQueue.offer(offset);
+        }
+        return offset;
+      }
+
+      if (timeStamp > targetTimestamp) {
+        // targetTimestamp is smaller than all existing messages's timestamp
+        if (offset - 1 < minOffset) {
+          if (offsetQueue.isEmpty()) {
+            offsetQueue.offer(minOffset);
+          }
+          return minOffset;
+        }
+        maxOffset = offset - 1;
+        long newOffset = minOffset + (maxOffset - minOffset) / 2;
+        LOG.debug("timeStamp > targetTimestamp, return newOffset {}", newOffset);
+        return newOffset;
+      }
+      if (timeStamp < targetTimestamp) {
+        // targetTimestamp is larger than all existing messages's timestamp, wait for new messages to come
+        if (offset + 1 > maxOffset) {
+          try {
+            Thread.sleep(1000);
+          } catch (InterruptedException e) {
+            LOG.info("Thread {} is interrupted while waiting for newer Kafka message.");
+            Thread.currentThread().interrupt();
+          }
+          maxOffset = getBoundaryOffset(LATEST) - 1;
+          return maxOffset;
+        }
+        minOffset = offset + 1;
+        long newOffset = minOffset + (maxOffset - minOffset) / 2;
```

Comment:
if the target timestamp is not in any of the log events, you will loop forever. When calculating the new offset, you need to make sure it's not the same as the current offset.